support/agent/sim.go

```diff
@@ -0,0 +1,251 @@
+package agent
+
+import (
+	"context"
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin/reward"
+	"math"
+	big2 "math/big"
+	"math/rand"
+	"strings"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/cbor"
+	"github.com/filecoin-project/go-state-types/exitcode"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin"
+	"github.com/filecoin-project/specs-actors/v2/actors/builtin/power"
+	vm "github.com/filecoin-project/specs-actors/v2/support/vm"
+)
+
+type Sim struct {
+	Config   SimConfig
+	Accounts []address.Address
+	Miners   []*MinerAgent
+	v        *vm.VM
+	rnd      *rand.Rand
+}
+
+type SimConfig struct {
+	AccountCount           int
+	AccountInitialBalance  abi.TokenAmount
+	Seed                   int64
+	CreateMinerProbability float32
+}
+
+type ReturnHandler func(v *vm.VM, msg Message, ret cbor.Marshaler) error
+
+type Message struct {
+	From          address.Address
+	To            address.Address
+	Value         abi.TokenAmount
+	Method        abi.MethodNum
+	Params        interface{}
+	ReturnHandler ReturnHandler
+}
+
+type MinerPowerTable struct {
+	addr    address.Address
+	qaPower abi.StoragePower
+}
+
+type PowerTable struct {
+	blockReward  abi.TokenAmount
+	totalQAPower abi.StoragePower
+	minerPower   []MinerPowerTable
+}
+
+func NewSim(ctx context.Context, t require.TestingT, config SimConfig) *Sim {
+	v := vm.NewVMWithSingletons(ctx, t)
+	return &Sim{
+		Config:   config,
+		Accounts: vm.CreateAccounts(ctx, t, v, config.AccountCount, config.AccountInitialBalance, config.Seed),
+		Miners:   []*MinerAgent{},
+		v:        v,
+		rnd:      rand.New(rand.NewSource(config.Seed)),
+	}
+}
+
+func (s *Sim) Tick() error {
+	var err error
+	var blockMessages []Message
+
+	// compute power table before state transition to create block rewards at the end
+	powerTable, err := ComputePowerTable(s.v, s.Miners)
+	if err != nil {
+		return err
+	}
+
+	// add all miner messages
+	for _, miner := range s.Miners {
```

Comment:
I think this is unnecessarily specialised to the concept of miners. Arbitrary agents could interact with the system. We don't have any right now, but it doesn't seem to cost anything to change the terminology here to allow for that.